File: app/models/qdm/device_request.rb
module QDM
  # app/models/qdm/device_request.rb
  class DeviceRequest < DataElement
    include Mongoid::Document
    embedded_in :patient
    field :identifier
    field :instantiatesCanonical
    field :instantiatesUri
    field :basedOn
    field :priorRequest
    field :groupIdentifier
    field :status
    field :intent
    field :priority
    field :code
    field :parameter
    field :subject
    field :encounter
    field :occurrence
    field :authoredOn
    field :requester
    field :performerType
    field :performer
    field :reasonQDM::Code
    field :reasonReference
    field :insurance
    field :supportingInfo
    field :note
    field :relevantHistory
    field :qdmVersion, type: String, default: '4.0.0'
  end
end

File: app/assets/javascripts/DeviceRequest.js
const mongoose = require('mongoose');
const { DataElementSchema } = require('./basetypes/DataElement');
const Code = require('./basetypes/Code');
const Interval = require('./basetypes/Interval');
const Quantity = require('./basetypes/Quantity');
const DateTime = require('./basetypes/DateTime');
const { ComponentSchema } = require('./Component');
const { FacilityLocationSchema } = require('./FacilityLocation');
const { IdSchema } = require('./Id');
const Any = require('./basetypes/Any');

const [Number, String] = [
  mongoose.Schema.Types.Number,
  mongoose.Schema.Types.String,
];

const DeviceRequestSchema = DataElementSchema({
  identifier: Any,
  instantiatesCanonical: Any,
  instantiatesUri: Any,
  basedOn: Any,
  priorRequest: Any,
  groupIdentifier: Any,
  status: Any,
  intent: Any,
  priority: Any,
  code: Any,
  parameter: Any,
  subject: Any,
  encounter: Any,
  occurrence: Any,
  authoredOn: Any,
  requester: Any,
  performerType: Any,
  performer: Any,
  reasonCode: Any,
  reasonReference: Any,
  insurance: Any,
  supportingInfo: Any,
  note: Any,
  relevantHistory: Any,
  qdmVersion: { type: String, default: '4.0.0' },
  _type: { type: String, default: 'DeviceRequest' },

});

module.exports.DeviceRequestSchema = DeviceRequestSchema;
module.exports.DeviceRequest = mongoose.model('DeviceRequest', DeviceRequestSchema);
